CurrentLocationKotlin/app/src/main/java/com/example/android/location/currentlocationkotlin/MainActivity.kt

```diff
@@ -15,12 +15,183 @@
+                            // Build intent that displays the App settings screen.
+                            val intent = Intent()
+                            intent.action = Settings.ACTION_APPLICATION_DETAILS_SETTINGS
+                            val uri = Uri.fromParts(
+                                "package",
+                                BuildConfig.APPLICATION_ID,
+                                null
+                            )
+                            intent.data = uri
+                            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
+                            startActivity(intent)
+                        }
+                        .show()
+                }
+            }
+        }
+    }
+
+    fun locationRequestOnClick(view: View) {
+        Log.d(TAG, "locationRequestOnClick()")
+
+        val permissionApproved =
+            applicationContext.hasPermission(Manifest.permission.ACCESS_FINE_LOCATION)
+
+        if (permissionApproved) {
+            requestCurrentLocation()
+        } else {
+            requestPermissionWithRationale(
+                Manifest.permission.ACCESS_FINE_LOCATION,
+                REQUEST_FINE_LOCATION_PERMISSIONS_REQUEST_CODE,
+                fineLocationRationalSnackbar
+            )
+        }
+    }
+
+    /**
+     * Gets current location.
+     * Note: The code checks for permission before calling this method, that is, it's never called
+     * from a method with a missing permission. Also, I include a second check with my extension
+     * function in case devs just copy/paste this code.
+     */
+    @SuppressLint("MissingPermission")
+    private fun requestCurrentLocation() {
+        Log.d(TAG, "requestCurrentLocation()")
+        if (applicationContext.hasPermission(Manifest.permission.ACCESS_FINE_LOCATION)) {
+
+            // Returns a single current location fix on the device. Unlike getLastLocation() that
+            // returns a cached location, this method could cause active location computation on the
+            // device. A single fresh location will be returned if the device location can be
+            // determined within reasonable time (tens of seconds), otherwise null will be returned.
+            //
+            // Both arguments are required.
+            // PRIORITY type is self-explanatory. (Other options are PRIORITY_BALANCED_POWER_ACCURACY,
+            // PRIORITY_LOW_POWER, and PRIORITY_NO_POWER.)
+            // The second parameter, [CancellationToken] allows the activity to cancel the request
+            // before completion.
+            val currentLocationTask: Task<Location> = fusedLocationClient.getCurrentLocation(
+                PRIORITY_HIGH_ACCURACY,
+                cancellationTokenSource.token
+            )
+
+            currentLocationTask.addOnCompleteListener { task: Task<Location> ->
+                val result = if (task.isSuccessful) {
+                    val result: Location = task.result
```

Comment:
The getCurentLocation can return a null result sometimes, is this handled here?

Reply:
I added a null check in the if statement. Let me know if that works.

Just to clarify, you are saying the task can return successful but the result will be null? If that is the case, should I add any comments about what that means or leave as it is now?

Reply:
Sometimes location cannot be computed, in that case a null location will be returned. It might be better to make the result nullable and check null explicitly.